augur/filters/exclude_by_nonnucleotide_symbols.py

```diff
@@ -1,53 +1,24 @@
-import functools
-
 from augur.filters.base_filter import BaseFilter
 
 
-def is_enabled(args):
-    return args.non_nucleotide
-
-
-def add_arguments(parser):
-    parser.add_argument(
-        "--non-nucleotide",
-        action="store_true",
-        help="exclude sequences that contain illegal characters",
-    )
-
-
-def init(context):
-    return ExcludeByNonnucleotideSymbols()
+class ExcludeByNonnucleotideSymbols(BaseFilter):
+    FILTER_OPERATION = "exclude"
 
+    def is_affected(self, sequence):
+        return sequence.has_only_nucleotide_symbols
```

Comment:
This is why I think FILTER_OPERATION and is_affected() should just be replaced with a boolean - 

in this case, FILTER_OPERATION is "exclude" - but `is_affected` returns True if the sequence has no illegal characters. So, is the sequence filtered out or not?

Compare with:
```python
class Filterer:
    <...>
    def filter_sequences(self):
        self.resulting_sequences = [
            sample for sample in self.sequences
            if self.apply_filter_chain(sample)
        ]

    def apply_filter_chain(self, sample):
        for filter_obj in self.filter_chain:
            if filter_obj.filter(sample) is False:
                return False
            return True

class NucleotideFilter:
    <...>
    def filter(sequence):
        return not sequence.has_only_nucleotide_symbols
```
(you can play with the negatives here, too - could do something like `filter.keep(sequence)` so we're not writing "not" everywhere)

Reply:
I was deferring the `"exclude"`/`"include"` resolution in the hope that there would be a generic solution to using `Subsample` as a filter (which might necessitate a third filter operation), but that just doesn't seem realistic.

If Subsample must be a separate thing, then we can use the fact that it is always applied after the exclusion filters and before the inclusion filters and just have two lists of filters.

It might look something like this (very rough example):

```python
class Filterer:
    def filter_sequences(self):
        sequences = [
            sample for sample in self.sequences
            if self.apply_exclusion_filter_chain(sample)
        ]
 
        sequences = self.subsample(sequences)

        sequences = [
            sample for sample in self.sequences  # note that this starts again with the original sequences
            if self.apply_inclusion_filter_chain(sample)
        ]

        self.resulting_sequences = sequences

    def apply_exclusion_filter_chain(self, sample):
        return any([filter_obj.is_affected(sample) for filter_obj in self.exclusion_filter_chain])

    def apply_inclusion_filter_chain(self, sample):
        # TODO maybe the boolean outcome if `is_affected` is flipped for these? eh.
        return any([filter_obj.is_affected(sample) for filter_obj in self.inclusion_filter_chain])
```

Reply:
Oh, I see, ok - that makes sense.

What do you think, then, about making that explicit and having a "before_subsample" class property?

```python

class BaseFilter:
    before_subsample = False #or whatever the default should be

class EarlyFilter(BaseFilter):
    before_subsample = True

class LateFilter(BaseFilter):
    <...>

class Filterer:
    def filter_sequences:
        sequences = self.filter_before_subsample() # applies EarlyFilter
        sequences = self.subsample(sequences)
        sequences = self.filter_after_subsample() # applies LateFilter
```
I'm just aiming for making it as explicit as possible what's happening and why we have the two separate lists - I understand the idea of having more "operations", but I think until we get there, my inclination is to be as simple and clear about what we're doing as possible.

Also, just so I understand - other than "include no matter what", what filters run after subsample? Is it possible to just skirt all this and hard-code one or two filters here?

Reply:
I just pushed some new changes.

- We've been circling around the idea of the "filters" just being dumb "matchers". Renamed to reflect this role.

- The matchers are so dumb that they don't even have a distinction between inclusion and exclusion. You were talking about this earlier. This means we can eliminate two of them: `ExcludeByName`/`IncludeByName` and `ExcludeByMetadata`/`IncludeByMetada` collapse.

- Whether a matcher is enabled comes straight from the command-line arguments—it's up to the user to _tell the program_ which matchers are meant for exclusion and which are meant for inclusion. Filterer has a translation function to support the legacy arguments. This means that any matcher can be used for inclusion or exclusion, for free!

- Renamed some other methods in Filterer, especially `no_exclusion_matches()` and `any_inclusion_matches()`.